types/artifacts.d.ts

```diff
@@ -743,6 +741,56 @@ declare global {
         /** Column number in the script (0-based). */
         columnNumber: number;
       }
+
+      /** Describes a generic console message. */
+      interface BaseConsoleMessage {
+        /**
+         * What emitted the console message (console API, violation log, etc.).
+         * See subtypes for possible values.
+         */
+        source: string,
+        /** The log level: info, warning, etc. See subtypes for possible values. */
+        level: string,
+        /**
+         * The text printed to the console, as shown on the browser. Note objects printed to the
+         * console will be stringified if primitive or JSON-serializable. Otherwise it will be
+         * a String representation of the object.
+         */
+        text: string;
+        /** Time of the console log in milliseconds since epoch. */
+        timestamp: number;
+        /** The stack trace of the log/exception, if known. */
+        stackTrace?: Crdp.Runtime.StackTrace;
+        /** The URL of the log/exception, if known. */
+        url?: string;
+        /** Line number in the script (0-based), if known. */
+        lineNumber?: number;
+        /** Column number in the script (0-based), if known. */
+        columnNumber?: number;
+      }
+
+      /** Describes a console message logged by a script using the console API. */
+      interface ConsoleAPICall extends BaseConsoleMessage {
+        /** The console API invoked. Only the following console API calls are gathered. */
+        source: 'console.warn' | 'console.error';
+        /** Corresponds to the API call. */
+        level: 'warning' | 'error';
```

Comment:
This seems redundant?

Reply:
Level definitely serves a different purpose from source, so definitely wouldn't want to eliminate one, but I agree with your main point. Maybe the `source` here just needs to be `console.log`|`console.*`|`console-js-api` or something rather than one for each? especially since we infer the specific source based on the level :)

Reply:
I definitely agree that it's redundant. My thought was that it's a nice way to document which API calls are gathered and which aren't. There are plenty of API calls besides `console.log`, `console.warn`, and `console.error` which are ignored, like `console.count`, `console.table`, etc.

Reply:
Distinguishing `.table`/`.dir`/etc was my initial thought as well @warrengm, but I changed my mind once I saw the impl inferred `console.warn` *from* the level which suggested we don't get the exact API info from the protocol.

Are you saying here that those APIs don't trigger any protocol event *at all*?

Reply:
It's actually getting the level from the `event.type` field, not `event.level`. I updated the code to clarify.

Oddly enough `console.warn` has `type` of `warning`: https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#event-consoleAPICalled

Reply:
Ooooh gotcha, I guess I should've jumped straight to protocol to check :)

Well then I'm back on team "let's be specific about the API being used" 🎉